File: src/stripe_entitlements/entitlements.py
from __future__ import annotations

import re
from collections.abc import Collection, Mapping
from dataclasses import dataclass
from datetime import datetime
from typing import Literal, cast

import asyncpg

from .bounds import JSON_SAFE_INTEGER_MAX
from .catalog import PlanCatalog
from .credit_amount import CREDIT_SCALE, CreditAmount
from .credits import (
    CreditDebitOwnerMismatchError,
    CreditResult,
    CreditService,
    CreditsUnavailableError,
    InsufficientCreditsError,
)
from .owner_reference import (
    InvalidOwnerReferenceError,
    validate_owner_external_ref,
)
from .subscription_state import (
    spendable_subscription_atoms,
    subscription_credits_are_spendable,
)

_ENTITLEMENT_KEY = re.compile(r"^[a-z][a-z0-9_]{0,63}$")

EntitlementReason = Literal[
    "allowed",
    "owner_not_found",
    "entitlement_not_enforceable",
    "feature_not_available",
    "limit_not_available",
    "limit_exceeded",
]
PlanInterval = Literal["month", "year"]
SubscriptionStatus = Literal["none", "active", "past_due", "canceled"]


class InvalidCreditRequestError(ValueError):
    pass


class BillingOwnerNotFoundError(LookupError):
    pass


class CreditOperationNotFoundError(LookupError):
    pass


class CreditIdempotencyConflictError(RuntimeError):
    pass


@dataclass(frozen=True, slots=True)
class LimitDecision:
    requested: int
    maximum: int | None
    allowed: bool


@dataclass(frozen=True, slots=True)
class EntitlementCheck:
    allowed: bool
    reason: EntitlementReason
    entitlements_enforceable: bool
    plan_key: str
    plan_interval: PlanInterval | None
    subscription_status: SubscriptionStatus
    credits_spendable: bool
    credit_balance: CreditAmount
    # Earliest expiry among funding that is currently spendable. This may be a
    # subscription window or a one-time credit-pack lot.
    credit_expires_at: datetime | None
    features: dict[str, bool]
    limits: dict[str, LimitDecision]

    @property
    def credit_scale(self) -> int:
        return CREDIT_SCALE


def _validate_operation_key(value: str) -> str:
    if (
        type(value) is not str
        or not value
        or value != value.strip()
        or len(value.encode("utf-8")) > 200
        or any(not character.isprintable() for character in value)
    ):
        raise InvalidCreditRequestError("idempotency key is invalid")
    return value


def _validate_requirements(
    required_features: Collection[str], required_limits: Mapping[str, int]
) -> tuple[tuple[str, ...], dict[str, int]]:
    features = tuple(required_features)
    if (
        len(features) > 64
        or len(features) != len(set(features))
        or any(
            type(feature) is not str or _ENTITLEMENT_KEY.fullmatch(feature) is None
            for feature in features
        )
    ):
        raise ValueError("required_features contains an invalid or duplicate key")
    limits = dict(required_limits)
    if len(limits) > 64 or any(
        type(key) is not str
        or _ENTITLEMENT_KEY.fullmatch(key) is None
        or type(value) is not int
        or value < 0
        or value > JSON_SAFE_INTEGER_MAX
        for key, value in limits.items()
    ):
        raise ValueError("required_limits contains an invalid key or value")
    return features, limits


class EntitlementService:
    """Server-side entitlement decisions and owner-bound credit operations."""

    def __init__(self, pool: asyncpg.Pool, catalog: PlanCatalog) -> None:
        self.pool = pool
        self.catalog = catalog
        self.credits = CreditService(pool)

    async def check(
        self,
        owner_external_ref: str,
        *,
        required_features: Collection[str] = (),
        required_limits: Mapping[str, int] | None = None,
    ) -> EntitlementCheck:
        owner_external_ref = validate_owner_external_ref(owner_external_ref)
        features, limits = _validate_requirements(required_features, required_limits or {})
        async with self.pool.acquire() as conn:
            account = await conn.fetchrow(
                """with observed as materialized (
                       select clock_timestamp() as database_now
                     )
                   select a.id,a.plan_key,a.plan_interval,a.subscription_status,
                          a.credits_balance,a.credit_expires_at,a.entitlement_revoked,
                          observed.database_now,
                          coalesce(p.pack_balance,0) as pack_balance,
                          p.next_pack_expiry
                     from billing_accounts a
                     cross join observed
                     left join lateral (
                         select sum(l.remaining_credits) as pack_balance,
                                min(l.expires_at) as next_pack_expiry
                           from credit_funding_lots l
                          where l.account_id=a.id
                            and l.status='active'
                            and l.remaining_credits > 0
                            and l.expires_at > observed.database_now
                     ) p on true
                    where a.external_ref=$1""",
                owner_external_ref,
            )
        if account is None:
            return EntitlementCheck(
                allowed=False,
                reason="owner_not_found",
                entitlements_enforceable=False,
                plan_key="free",
                plan_interval=None,
                subscription_status="none",
                credits_spendable=False,
                credit_balance=CreditAmount.from_atoms(0),
                credit_expires_at=None,
                features={feature: False for feature in features},
                limits={
                    key: LimitDecision(requested=value, maximum=None, allowed=False)
                    for key, value in limits.items()
                },
            )

        plan_key = str(account["plan_key"])
        plan = self.catalog.plans.get(plan_key)
        database_now = account["database_now"]
        expires_at = account["credit_expires_at"]
        subscription_spendable = subscription_credits_are_spendable(account, as_of=database_now)
        enforceable = bool(plan is not None and subscription_spendable)
        subscription_atoms = spendable_subscription_atoms(account, as_of=database_now)
        pack_atoms = int(account["pack_balance"])
        spendable_atoms = subscription_atoms + pack_atoms
        funding_expiries = []
        if subscription_atoms > 0 and isinstance(expires_at, datetime):
            funding_expiries.append(expires_at)
        next_pack_expiry = account["next_pack_expiry"]
        if pack_atoms > 0 and isinstance(next_pack_expiry, datetime):
            funding_expiries.append(next_pack_expiry)
        next_funding_expiry = min(funding_expiries) if funding_expiries else None
        feature_decisions = {
            feature: bool(enforceable and plan is not None and feature in plan.features)
            for feature in features
        }
        limit_decisions: dict[str, LimitDecision] = {}
        for key, requested in limits.items():
            maximum = plan.limits.get(key) if plan is not None else None
            limit_decisions[key] = LimitDecision(
                requested=requested,
                maximum=maximum,
                allowed=bool(enforceable and maximum is not None and requested <= maximum),
            )

        reason: EntitlementReason = "allowed"
        allowed = enforceable
        if not enforceable:
            reason = "entitlement_not_enforceable"
        elif not all(feature_decisions.values()):
            reason = "feature_not_available"
            allowed = False
        elif any(decision.maximum is None for decision in limit_decisions.values()):
            reason = "limit_not_available"
            allowed = False
        elif not all(decision.allowed for decision in limit_decisions.values()):
            reason = "limit_exceeded"
            allowed = False

        return EntitlementCheck(
            allowed=allowed,
            reason=reason,
            entitlements_enforceable=enforceable,
            plan_key=plan_key,
            plan_interval=cast(PlanInterval | None, account["plan_interval"]),
            subscription_status=cast(SubscriptionStatus, account["subscription_status"]),
            credits_spendable=spendable_atoms > 0,
            credit_balance=CreditAmount.from_atoms(spendable_atoms),
            credit_expires_at=next_funding_expiry,
            features=feature_decisions,
            limits=limit_decisions,
        )

    async def charge(
        self,
        owner_external_ref: str,
        amount: str,
        idempotency_key: str,
    ) -> CreditResult:
        owner_external_ref = validate_owner_external_ref(owner_external_ref)
        idempotency_key = _validate_operation_key(idempotency_key)
        if type(amount) is not str:
            raise InvalidCreditRequestError("credit amount must be an exact decimal string")
        try:
            normalized = CreditAmount.parse(amount, field="credit amount", allow_zero=False)
        except ValueError as exc:
            raise InvalidCreditRequestError("credit amount is invalid") from exc
        async with self.pool.acquire() as conn:
            account_id = await conn.fetchval(
                "select id from billing_accounts where external_ref=$1",
                owner_external_ref,
            )
        if account_id is None:
            raise BillingOwnerNotFoundError("billing owner not found")
        try:
            return await self.credits.charge(str(account_id), normalized, idempotency_key)
        except KeyError as exc:
            raise BillingOwnerNotFoundError("billing owner not found") from exc
        except ValueError as exc:
            raise CreditIdempotencyConflictError("idempotency key conflict") from exc

    async def refund(self, owner_external_ref: str, idempotency_key: str) -> CreditResult:
        owner_external_ref = validate_owner_external_ref(owner_external_ref)
        idempotency_key = _validate_operation_key(idempotency_key)
        async with self.pool.acquire() as conn:
            account_id = await conn.fetchval(
                "select id from billing_accounts where external_ref=$1",
                owner_external_ref,
            )
        if account_id is None:
            raise CreditOperationNotFoundError("credit operation not found")
        try:
            return await self.credits.refund(
                idempotency_key,
                expected_account_id=str(account_id),
            )
        except (CreditDebitOwnerMismatchError, KeyError) as exc:
            raise CreditOperationNotFoundError("credit operation not found") from exc


__all__ = [
    "BillingOwnerNotFoundError",
    "CreditIdempotencyConflictError",
    "CreditOperationNotFoundError",
    "CreditsUnavailableError",
    "EntitlementCheck",
    "EntitlementReason",
    "EntitlementService",
    "InsufficientCreditsError",
    "InvalidCreditRequestError",
    "InvalidOwnerReferenceError",
    "LimitDecision",
    "PlanInterval",
    "SubscriptionStatus",
    "validate_owner_external_ref",
]

File: typescript/src/entitlements.ts
import type { QueryResultRow } from "pg";

import { JSON_SAFE_INTEGER_MAX } from "./bounds.js";
import type { PlanCatalog } from "./catalog.js";
import { CREDIT_SCALE, CreditAmount } from "./credit-amount.js";
import {
  CreditAccountNotFoundError,
  CreditDebitNotFoundError,
  CreditDebitOwnerMismatchError,
  CreditService,
  CreditsUnavailableError,
  InsufficientCreditsError,
} from "./credits.js";
import type { CreditResult } from "./credits.js";
import type { Database } from "./database.js";
import { pgBigInt } from "./db-types.js";
import {
  spendableSubscriptionAtoms,
  subscriptionCreditsAreSpendable,
} from "./subscription-state.js";
import type { BillingInterval, PgTimestamp } from "./types.js";
import {
  InvalidOwnerReferenceError,
  validateOwnerExternalRef,
} from "./owner-reference.js";
import { isPlainRecord, isPrintable } from "./validation.js";

const ENTITLEMENT_KEY = /^[a-z][a-z0-9_]{0,63}$/u;

export type EntitlementReason =
  | "allowed"
  | "owner_not_found"
  | "entitlement_not_enforceable"
  | "feature_not_available"
  | "limit_not_available"
  | "limit_exceeded";

export type SubscriptionStatus = "none" | "active" | "past_due" | "canceled";

export class InvalidCreditRequestError extends Error {}
export class BillingOwnerNotFoundError extends Error {}
export class CreditOperationNotFoundError extends Error {}
export class CreditIdempotencyConflictError extends Error {}

export interface LimitDecision {
  readonly requested: number;
  readonly maximum: number | null;
  readonly allowed: boolean;
}

export class EntitlementCheck {
  public readonly allowed: boolean;
  public readonly reason: EntitlementReason;
  public readonly entitlementsEnforceable: boolean;
  public readonly planKey: string;
  public readonly planInterval: BillingInterval | null;
  public readonly subscriptionStatus: SubscriptionStatus;
  public readonly creditsSpendable: boolean;
  public readonly creditBalance: CreditAmount;
  public readonly creditExpiresAt: PgTimestamp | null;
  public readonly features: Readonly<Record<string, boolean>>;
  public readonly limits: Readonly<Record<string, LimitDecision>>;

  public constructor(input: {
    readonly allowed: boolean;
    readonly reason: EntitlementReason;
    readonly entitlementsEnforceable: boolean;
    readonly planKey: string;
    readonly planInterval: BillingInterval | null;
    readonly subscriptionStatus: SubscriptionStatus;
    readonly creditsSpendable: boolean;
    readonly creditBalance: CreditAmount;
    readonly creditExpiresAt: PgTimestamp | null;
    readonly features: Readonly<Record<string, boolean>>;
    readonly limits: Readonly<Record<string, LimitDecision>>;
  }) {
    this.allowed = input.allowed;
    this.reason = input.reason;
    this.entitlementsEnforceable = input.entitlementsEnforceable;
    this.planKey = input.planKey;
    this.planInterval = input.planInterval;
    this.subscriptionStatus = input.subscriptionStatus;
    this.creditsSpendable = input.creditsSpendable;
    this.creditBalance = input.creditBalance;
    this.creditExpiresAt = input.creditExpiresAt;
    this.features = input.features;
    this.limits = input.limits;
  }

  public get creditScale(): bigint {
    return CREDIT_SCALE;
  }
}

interface EntitlementAccountRow extends QueryResultRow {
  readonly id: string;
  readonly plan_key: string;
  readonly plan_interval: BillingInterval | null;
  readonly subscription_status: SubscriptionStatus;
  readonly credits_balance: string;
  readonly credit_expires_at: PgTimestamp | null;
  readonly entitlement_revoked: boolean;
  readonly database_now: PgTimestamp;
  readonly pack_balance: string;
  readonly next_pack_expiry: PgTimestamp | null;
  readonly combined_expiry: PgTimestamp | null;
}

function validateOperationKey(value: string): string {
  if (
    typeof value !== "string" ||
    value.length === 0 ||
    value !== value.trim() ||
    Buffer.byteLength(value, "utf8") > 200 ||
    !isPrintable(value)
  ) {
    throw new InvalidCreditRequestError("idempotency key is invalid");
  }
  return value;
}

function validateRequirements(
  requiredFeatures: readonly string[] | ReadonlySet<string>,
  requiredLimits: Readonly<Record<string, number>>,
): readonly [readonly string[], Readonly<Record<string, number>>] {
  const features = [...requiredFeatures];
  if (
    features.length > 64 ||
    features.length !== new Set(features).size ||
    features.some(
      (feature) =>
        typeof feature !== "string" || !ENTITLEMENT_KEY.test(feature),
    )
  ) {
    throw new TypeError(
      "requiredFeatures contains an invalid or duplicate key",
    );
  }
  if (!isPlainRecord(requiredLimits)) {
    throw new TypeError("requiredLimits must be a plain record");
  }
  const limits = { ...requiredLimits };
  const entries = Object.entries(limits);
  if (
    entries.length > 64 ||
    entries.some(
      ([key, value]) =>
        !ENTITLEMENT_KEY.test(key) ||
        !Number.isSafeInteger(value) ||
        value < 0 ||
        value > JSON_SAFE_INTEGER_MAX,
    )
  ) {
    throw new TypeError("requiredLimits contains an invalid key or value");
  }
  return [features, limits];
}

function accountView(
  account: EntitlementAccountRow,
): Readonly<Record<string, unknown>> {
  return {
    ...account,
    credits_balance: pgBigInt(
      account.credits_balance,
      "account credits_balance",
    ),
  };
}

/** Server-side entitlement decisions and owner-bound credit operations. */
export class EntitlementService {
  readonly #database: Database;
  readonly #catalog: PlanCatalog;
  readonly #credits: CreditService;

  public constructor(database: Database, catalog: PlanCatalog) {
    this.#database = database;
    this.#catalog = catalog;
    this.#credits = new CreditService(database);
  }

  public async check(
    ownerExternalRefInput: string,
    options: {
      readonly requiredFeatures?: readonly string[] | ReadonlySet<string>;
      readonly requiredLimits?: Readonly<Record<string, number>>;
    } = {},
  ): Promise<EntitlementCheck> {
    const ownerExternalRef = validateOwnerExternalRef(ownerExternalRefInput);
    const [features, limits] = validateRequirements(
      options.requiredFeatures ?? [],
      options.requiredLimits ?? {},
    );
    const accountResult = await this.#database.query<EntitlementAccountRow>(
      `with observed as materialized (
         select clock_timestamp() as database_now
       )
       select a.id,a.plan_key,a.plan_interval,a.subscription_status,
              a.credits_balance,a.credit_expires_at,a.entitlement_revoked,
              observed.database_now,
              coalesce(p.pack_balance,0) as pack_balance,
              p.next_pack_expiry,
              least(a.credit_expires_at,p.next_pack_expiry) as combined_expiry
         from billing_accounts a
         cross join observed
         left join lateral (
           select sum(l.remaining_credits) as pack_balance,
                  min(l.expires_at) as next_pack_expiry
             from credit_funding_lots l
            where l.account_id=a.id and l.status='active'
              and l.remaining_credits > 0
              and l.expires_at > observed.database_now
         ) p on true
        where a.external_ref=$1`,
      [ownerExternalRef],
    );
    const account = accountResult.rows[0];
    if (account === undefined) {
      return new EntitlementCheck({
        allowed: false,
        reason: "owner_not_found",
        entitlementsEnforceable: false,
        planKey: "free",
        planInterval: null,
        subscriptionStatus: "none",
        creditsSpendable: false,
        creditBalance: CreditAmount.fromAtoms(0n),
        creditExpiresAt: null,
        features: Object.fromEntries(
          features.map((feature) => [feature, false]),
        ),
        limits: Object.fromEntries(
          Object.entries(limits).map(([key, requested]) => [
            key,
            { requested, maximum: null, allowed: false },
          ]),
        ),
      });
    }

    const plan = this.#catalog.plans.get(account.plan_key);
    const view = accountView(account);
    const subscriptionSpendable = subscriptionCreditsAreSpendable(view, {
      asOf: account.database_now,
    });
    const enforceable = plan !== undefined && subscriptionSpendable;
    const subscriptionAtoms = spendableSubscriptionAtoms(view, {
      asOf: account.database_now,
    });
    const packAtoms = pgBigInt(account.pack_balance, "pack credit balance");
    const spendableAtoms = subscriptionAtoms + packAtoms;
    let nextFundingExpiry: PgTimestamp | null = null;
    if (subscriptionAtoms > 0n && packAtoms > 0n) {
      nextFundingExpiry = account.combined_expiry;
    } else if (subscriptionAtoms > 0n) {
      nextFundingExpiry = account.credit_expires_at;
    } else if (packAtoms > 0n) {
      nextFundingExpiry = account.next_pack_expiry;
    }

    const featureDecisions = Object.fromEntries(
      features.map((feature) => [
        feature,
        enforceable && plan.features.has(feature),
      ]),
    );
    const limitDecisions: Record<string, LimitDecision> = {};
    for (const [key, requested] of Object.entries(limits)) {
      const maximum = plan?.limits[key] ?? null;
      limitDecisions[key] = {
        requested,
        maximum,
        allowed: enforceable && maximum !== null && requested <= maximum,
      };
    }

    let reason: EntitlementReason = "allowed";
    let allowed = enforceable;
    if (!enforceable) {
      reason = "entitlement_not_enforceable";
    } else if (Object.values(featureDecisions).some((value) => !value)) {
      reason = "feature_not_available";
      allowed = false;
    } else if (
      Object.values(limitDecisions).some(
        (decision) => decision.maximum === null,
      )
    ) {
      reason = "limit_not_available";
      allowed = false;
    } else if (
      Object.values(limitDecisions).some((decision) => !decision.allowed)
    ) {
      reason = "limit_exceeded";
      allowed = false;
    }

    return new EntitlementCheck({
      allowed,
      reason,
      entitlementsEnforceable: enforceable,
      planKey: account.plan_key,
      planInterval: account.plan_interval,
      subscriptionStatus: account.subscription_status,
      creditsSpendable: spendableAtoms > 0n,
      creditBalance: CreditAmount.fromAtoms(spendableAtoms),
      creditExpiresAt: nextFundingExpiry,
      features: featureDecisions,
      limits: limitDecisions,
    });
  }

  public async charge(
    ownerExternalRefInput: string,
    amount: string,
    idempotencyKeyInput: string,
  ): Promise<CreditResult> {
    const ownerExternalRef = validateOwnerExternalRef(ownerExternalRefInput);
    const idempotencyKey = validateOperationKey(idempotencyKeyInput);
    if (typeof amount !== "string") {
      throw new InvalidCreditRequestError(
        "credit amount must be an exact decimal string",
      );
    }
    let normalized: CreditAmount;
    try {
      normalized = CreditAmount.parse(amount, {
        field: "credit amount",
        allowZero: false,
      });
    } catch (error) {
      throw new InvalidCreditRequestError("credit amount is invalid", {
        cause: error,
      });
    }
    const accountResult = await this.#database.query<
      { readonly id: string } & QueryResultRow
    >("select id from billing_accounts where external_ref=$1", [
      ownerExternalRef,
    ]);
    const accountId = accountResult.rows[0]?.id;
    if (accountId === undefined) {
      throw new BillingOwnerNotFoundError("billing owner not found");
    }
    try {
      return await this.#credits.charge(accountId, normalized, idempotencyKey);
    } catch (error) {
      if (error instanceof CreditAccountNotFoundError) {
        throw new BillingOwnerNotFoundError("billing owner not found", {
          cause: error,
        });
      }
      if (
        error instanceof TypeError &&
        error.message ===
          "idempotency key was already used with different parameters"
      ) {
        throw new CreditIdempotencyConflictError("idempotency key conflict", {
          cause: error,
        });
      }
      throw error;
    }
  }

  public async refund(
    ownerExternalRefInput: string,
    idempotencyKeyInput: string,
  ): Promise<CreditResult> {
    const ownerExternalRef = validateOwnerExternalRef(ownerExternalRefInput);
    const idempotencyKey = validateOperationKey(idempotencyKeyInput);
    const accountResult = await this.#database.query<
      { readonly id: string } & QueryResultRow
    >("select id from billing_accounts where external_ref=$1", [
      ownerExternalRef,
    ]);
    const accountId = accountResult.rows[0]?.id;
    if (accountId === undefined) {
      throw new CreditOperationNotFoundError("credit operation not found");
    }
    try {
      return await this.#credits.refund(idempotencyKey, {
        expectedAccountId: accountId,
      });
    } catch (error) {
      if (
        error instanceof CreditDebitOwnerMismatchError ||
        error instanceof CreditDebitNotFoundError ||
        error instanceof CreditAccountNotFoundError
      ) {
        throw new CreditOperationNotFoundError("credit operation not found", {
          cause: error,
        });
      }
      throw error;
    }
  }
}

export {
  CreditsUnavailableError,
  InsufficientCreditsError,
  InvalidOwnerReferenceError,
  validateOwnerExternalRef,
};
